traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go

```diff
@@ -450,6 +472,177 @@ func createV31(w http.ResponseWriter, r *http.Request, inf *api.APIInfo, dsV31 t
 	return &dsV31, http.StatusOK, nil, nil
 }
 
+func createV40(w http.ResponseWriter, r *http.Request, inf *api.APIInfo, dsV40 tc.DeliveryServiceV40) (*tc.DeliveryServiceV40, int, error, error) {
```

Comment:
We shouldn't have to create another copy of this function. Instead, the `createV31` function should call _this_ function, take the result, and convert it back to a V31 struct to return back to the client.

In `createV40` we don't write `CacheURL` to the DB, so that should instead be handled by the `createV31` function after getting a successful return value from `createV40`. If `EnsureParams` handles CacheURL-ey stuff as well, it would go in `createV31` alone.

Also, `createV40` should cast the `tc.DeliveryServiceV40` into the `tc.DeliveryServiceV4` alias at the beginning of the method then cast back to `tc.DeliveryServiceV40` at the end (on L479 and L642, respectively) in order for all the other functions used by this function to accept the `tc.DeliveryServiceV4` alias as input. Otherwise, every new API version requires updating all of these extra functions.